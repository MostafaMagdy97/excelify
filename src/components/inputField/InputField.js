// @flow

import React from 'react';
import { Box } from 'grid-styled';
import styled from 'styled-components';
import { fontSize, borderColor } from 'styled-system';

import Text from '../text/Text';
import Icon from '../icon/Icon';

import { FONT_TYPES, FONT_WEIGHTS, FONT_SIZES } from '../../base/Typography';
import { COLORS_VALUES, COLORS } from '../../base/Colors';
import Space from '../../base/Space';

const InputFieldContainer = styled.div`
  border: 1px solid;
  ${borderColor};
  height: 38px;
  width: ${props => (props.width ? props.width : 'max-content')};
  display: flex;
  align-items: center;
  padding: 0 ${Space[3]}px;
  border-radius: 4px;
`;

const StyledInput = styled.input`
  border: none;
  outline: none;
  ${fontSize};
  font-weight: ${FONT_WEIGHTS.normal};
  color: ${COLORS_VALUES[COLORS.TEXT]};
  width: 100%;
`;

const ErrorMessage = styled(Text)`
  height: 24px;
  display: flex;
  align-items: center;
  width: ${props => props.width};
`;

type Props = {
  type: string,
  width: number | string,
  icon: Object,
  iconWidth: number,
  placeholder: string,
  value: ?string,
  onChange: Function,
  onBlur: Function,
  onFocus: Function,
  isRequired: boolean,
  disable: boolean,
  callbackParams: any,
  regexArray: Object,
  hideErrorMessage: boolean,
};

type State = {
  color: string,
  errorMessage: ?string,
};

class InputField extends React.Component<Props, State> {
  static defaultProps = {
    type: 'string',
    hideErrorMessage: false,
    onChange: () => {},
    onBlur: () => {},
    onFocus: () => {},
  };

  state = {
    color: COLORS.TEXT,
    errorMessage: undefined,
  };

  onFocus = () => {
    const { onFocus } = this.props;
    this.hideErrorMessage();

    this.setState(() => {
      onFocus();

      return {
        color: COLORS.BLUE,
      };
    });
  };

  onBlur = () => {
    const { onBlur } = this.props;
    const { isValid, errorMessage } = this.isValid();

    this.setState(() => {
      if (isValid) {
        onBlur();

        return {
          color: COLORS.TEXT,
        };
      }

      this.showErrorMessage(errorMessage);
      return null;
    });
  };

  onChange = (e: Object) => {
    const { callbackParams } = this.props;
    const { value } = e.target;
    this.props.onChange(value, callbackParams);
  };

  isValid = () => {
    const { disable, value, placeholder, isRequired, regexArray } = this.props;
    let shouldBreak = false;
    let result = {
      isValid: true,
      errorMessage: undefined,
    };

    if (!disable) {
      if (value === '' || value === undefined) {
        if (isRequired === true) {
          result = {
            isValid: false,
            errorMessage: `${placeholder} is required`,
          };
        }
      } else if (regexArray) {
        regexArray.forEach(check => {
          if (shouldBreak) return;
          const regexResult = check.regex.test(value);

          if (!regexResult) {
            shouldBreak = true;
            result = {
              isValid: false,
              errorMessage: check.errorMessage,
            };
          } else {
            result = {
              isValid: true,
              errorMessage: undefined,
            };
          }
        });
      } else {
        result = {
          isValid: true,
          errorMessage: undefined,
        };
      }
    }

    return result;
  };

  showErrorMessage = (errorMessage: ?string) => {
    this.setState(() => ({
      errorMessage,
      color: COLORS.DANGER,
    }));
  };

  hideErrorMessage = () => {
    this.setState({
      errorMessage: undefined,
      color: COLORS.TEXT,
    });
  };

  render() {
    const {
      icon,
      iconWidth,
      placeholder,
      value,
      disable,
      type,
      width,
      hideErrorMessage,
    } = this.props;
    const { color, errorMessage } = this.state;

    return (
      <div>
        <InputFieldContainer width={width} borderColor={color}>
          {icon && (
            <React.Fragment>
              <Icon icon={icon} width={iconWidth} color={color} />
              <Box mr={Space[3]} />
            </React.Fragment>
          )}
          <StyledInput
            placeholder={placeholder}
            value={value}
            onFocus={this.onFocus}
            onBlur={this.onBlur}
            onChange={this.onChange}
            disabled={disable}
            type={type}
            fontSize={FONT_SIZES[FONT_TYPES.BODY]}
          />
        </InputFieldContainer>
        {!hideErrorMessage && (
          <ErrorMessage type={FONT_TYPES.CAPTION} color={COLORS.DANGER} width={width}>
            {errorMessage}
          </ErrorMessage>
        )}
      </div>
    );
  }
}

export default InputField;
