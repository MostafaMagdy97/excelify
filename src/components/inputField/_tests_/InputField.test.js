import React from 'react';
import { configure, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import 'jest-enzyme';
import 'jest-styled-components';

import InputField from '../InputField';

configure({ adapter: new Adapter() });

describe('InputField', () => {
  it('should render', () => {
    const tree = render(<InputField placeholder="test" />);
    expect(tree).toMatchSnapshot();
  });
});
