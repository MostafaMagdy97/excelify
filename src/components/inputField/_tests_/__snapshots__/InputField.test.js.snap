// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField should render 1`] = `
.c3 {
  color: danger;
  font-weight: 500;
  font-size: 13px;
  font-display: fallback;
}

.c0 {
  border: 1px solid;
  border-color: text;
  height: 38px;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 12px;
  border-radius: 4px;
}

.c1 {
  border: none;
  outline: none;
  font-size: 15px;
  color: #1b2733;
  width: 100%;
}

.c2 {
  height: 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

@media screen and (min-width:40em) {
  .c3 {
    font-size: 13px;
  }
}

@media screen and (min-width:52em) {
  .c3 {
    font-size: 12px;
  }
}

@media screen and (min-width:64em) {
  .c3 {
    font-size: 12px;
  }
}

@media screen and (min-width:40em) {
  .c1 {
    font-size: 15px;
  }
}

@media screen and (min-width:52em) {
  .c1 {
    font-size: 14px;
  }
}

@media screen and (min-width:64em) {
  .c1 {
    font-size: 14px;
  }
}

<div>
  <div
    class="c0"
  >
    <input
      class="c1"
      font-size="15px,15px,14px,14px"
      placeholder="test"
      type="string"
    >
      
    </input>
  </div>
  <label
    class="c2 c3"
    color="danger"
    font-size="13px,13px,12px,12px"
    font-weight="500"
    tag="label"
    type="caption"
  >
    
  </label>
</div>
`;
